service/history/shard/context.go

```diff
@@ -176,16 +176,21 @@ var _ Context = (*contextImpl)(nil)
 var (
 	// ErrShardClosed is returned when shard is closed and a req cannot be processed
 	ErrShardClosed = errors.New("shard closed")
+
+	// ErrShardRecentlyClosed is returned when a shard has recently been closed, this
+	// error is transient, and should not cause error logs and error metrics to be emitted
+	ErrShardRecentlyClosed = errors.New("shard recently closed")
 )
 
 const (
 	// transfer/cross cluster diff/lag is in terms of taskID, which is calculated based on shard rangeID
 	// on shard movement, taskID will increase by around 1 million
-	logWarnTransferLevelDiff    = 3000000 // 3 million
-	logWarnCrossClusterLevelLag = 3000000 // 3 million
-	logWarnTimerLevelDiff       = time.Duration(30 * time.Minute)
-	historySizeLogThreshold     = 10 * 1024 * 1024
-	minContextTimeout           = 1 * time.Second
+	logWarnTransferLevelDiff     = 3000000 // 3 million
+	logWarnCrossClusterLevelLag  = 3000000 // 3 million
+	logWarnTimerLevelDiff        = time.Duration(30 * time.Minute)
+	historySizeLogThreshold      = 10 * 1024 * 1024
+	minContextTimeout            = 1 * time.Second
+	timeBeforeShardClosedIsError = 10 * time.Second
```

Comment:
did you found cases shard is closed momentarily and and re-opened? In which cases this can happen?

Reply:
We have seen cases where the same host keeps emitting the `shard closed` error for good. We are not sure what causes this, but we have seen it happen. We then have had to restart the history host. 

This is not root caused, so we should still have these errors emitted if we keep seeing this a long time after the shard was closed.

Reply:
I investigated that nonstop `shard closed` errors issue a few weeks ago and root cause is the [failover queues that are created but never stopped](https://github.com/uber/cadence/blob/2f67371c07ee68d157ffd1afb1a072f9c93ea18e/service/history/queue/transfer_queue_processor.go#L290). See CDNC-8135